src/Storages/StorageS3.h

```diff
@@ -60,6 +60,10 @@ class StorageS3Source : public ISource, WithContext
         virtual ~IIterator() = default;
         virtual KeyWithInfo next() = 0;
 
+        /// Estimates how many streams we need to process all files.
+        /// If keys count >= max_threads_count, the returned number may not represent the actual number of the keys.
```

Comment:
Nitpick: there are more caveats:
 * Intended to be called before any `next()` calls, may underestimate otherwise.
 * May underestimate if the glob has a strong filter, so there are few matches among the first 1000 ListObjects results.